File: src/main/java/com/derongan/minecraft/guiy/gui/Element.java
package com.derongan.minecraft.guiy.gui;

/**
 * Interface shared by all GUI elements.
 */
public interface Element {
    /**
     * Get the size of the element.
     */
    Size getSize();

    /**
     * Draws the element to the provided {@link GuiRenderer}.
     */
    void draw(GuiRenderer guiRenderer);

    /**
     * Runs when an element is clicked. This generally includes any slot that the element covers with its Size.
     */
    default void onClick(ClickEvent clickEvent) {
        clickEvent.setCancelled(true);
    }
}

File: src/main/java/com/derongan/minecraft/guiy/gui/Element.kt
package com.derongan.minecraft.guiy.gui

import com.derongan.minecraft.guiy.kotlin_dsl.GuiyMarker

/**
 * Interface shared by all GUI elements.
 *
 * @property dims The [Size] of the element.
 */
@GuiyMarker
interface Element {
    val dims: Size
//    val constraints: Constraints get() = Constraints()

    /**
     * Draws the element to the provided [GuiRenderer].
     */
    fun draw(guiRenderer: GuiRenderer)

    /**
     * Runs when an element is clicked. This generally includes any slot that the element covers with its Size.
     */
    fun onClick(clickEvent: ClickEvent) {
        clickEvent.isCancelled = true
    }
}